crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -108,57 +185,135 @@ impl<S: States> Deref for State<S> {
     }
 }
 
+#[derive(Resource, Default, Debug)]
+#[cfg_attr(
+    feature = "bevy_reflect",
+    derive(bevy_reflect::Reflect),
+    reflect(Resource, Default)
+)]
+pub(crate) struct ActiveTransition<S: States>(Option<S>, Option<S>);
+
+impl<S: States> ActiveTransition<S> {
+    pub(crate) fn new(main: Option<S>, secondary: Option<S>) -> Self {
+        Self(main, secondary)
+    }
+
+    pub(crate) fn swap(&mut self) {
+        let main = self.0.clone();
+        self.0 = self.1.clone();
+        self.1 = main;
+    }
+
+    pub(crate) fn get_main(&self) -> Option<&S> {
+        self.0.as_ref()
+    }
+
+    pub(crate) fn get_secondary(&self) -> Option<&S> {
+        self.1.as_ref()
+    }
+}
+
 /// The next state of [`State<S>`].
 ///
 /// To queue a transition, just set the contained value to `Some(next_state)`.
 /// Note that these transitions can be overridden by other systems:
 /// only the actual value of this resource at the time of [`apply_state_transition`] matters.
-#[derive(Resource, Default, Debug)]
+#[derive(Resource, Default)]
 #[cfg_attr(
     feature = "bevy_reflect",
     derive(bevy_reflect::Reflect),
     reflect(Resource, Default)
 )]
-pub struct NextState<S: States>(pub Option<S>);
+pub enum NextState<S: States> {
+    /// Do not change the state.
+    #[default]
+    Keep,
+    /// Change the state to a specific, pre-determined value
+    Value(S),
+    /// Change the state to a value determined by the given closure
+    Setter(
+        #[cfg_attr(feature = "bevy_reflect", reflect(ignore, default = "default_setter"))]
+        Box<dyn Fn(S) -> S + Sync + Send>,
+    ),
+}
+
+fn default_setter<S: States>() -> Box<dyn Fn(S) -> S + Sync + Send> {
+    Box::new(|state: S| state)
+}
+
+impl<S: States> Debug for NextState<S> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Keep => write!(f, "Keep"),
+            Self::Value(arg0) => f.debug_tuple("Value").field(arg0).finish(),
+            Self::Setter(_) => write!(f, "Setter"),
+        }
+    }
+}
 
 impl<S: States> NextState<S> {
     /// Tentatively set a planned state transition to `Some(state)`.
     pub fn set(&mut self, state: S) {
-        self.0 = Some(state);
+        *self = Self::Value(state);
+    }
+    /// Tentatively set a planned state transition to `Some(state)`.
+    pub fn setter(&mut self, setter: impl Fn(S) -> S + 'static + Sync + Send) {
```

Comment:
Not the clearest name or usage. A doc test might be helpful here.

Reply:
Added a doctest